### PR TITLE
Expose external refreshes through the stats API

### PR DESCRIPTION
**Pull request by clandry94**
An implementation from this issue https://github.com/elastic/elasticsearch/issues/36712.

Right now, the stats API only provides refresh metrics regarding internal refreshes. This isn't very useful and somewhat misleading for cluster administrators since internal refreshes are not indicative of documents being available for search..

In this PR I added a new metric for collecting external refreshes as they occur and exposing them through the stats API. Now, calling an endpoint for stats will yield external refresh metrics as well.
```
GET /test/_stats/refresh
```
```
{
  "_shards" : {
    "total" : 2,
    "successful" : 1,
    "failed" : 0
  },
  "_all" : {
    "primaries" : {
      "refresh" : {
        "total" : 7,
        "total_time_in_millis" : 157,
        "external_total" : 5,
        "external_total_time_in_millis" : 157,
        "listeners" : 0
      }
    },
    "total" : {
      "refresh" : {
        "total" : 7,
        "total_time_in_millis" : 157,
        "external_total" : 5,
        "external_total_time_in_millis" : 157,
        "listeners" : 0
      }
    }
  },
  "indices" : {
    "test" : {
      "uuid" : "4UxNKfIjSCGvGrBEU1l7tg",
      "primaries" : {
        "refresh" : {
          "total" : 7,
          "total_time_in_millis" : 157,
          "external_total" : 5,
          "external_total_time_in_millis" : 157,
          "listeners" : 0
        }
      },
      "total" : {
        "refresh" : {
          "total" : 7,
          "total_time_in_millis" : 157,
          "external_total" : 5,
          "external_total_time_in_millis" : 157,
          "listeners" : 0
        }
      }
    }
  }
}
```

Also, how is the naming? I'm not sure `external_total` and `external_total_time_in_millis` is the best way to express this, but it gets the point across. Also, should `total` and `total_time_in_millis` be changed to `internal_total` and so on? 

cc @s1monw

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -214,7 +214,10 @@ Runnable getGlobalCheckpointSyncer() {
     private RecoveryState recoveryState;
 
     private final RecoveryStats recoveryStats = new RecoveryStats();
+
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: extra newline

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -214,7 +214,10 @@ Runnable getGlobalCheckpointSyncer() {
...
     private final MeanMetric refreshMetric = new MeanMetric();
+    private final MeanMetric externalRefreshMetric = new MeanMetric();
+
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: extra newline

**Review by s1monw — APPROVED** [review]
LGTM left some nit-picks. I am sorry for the delay on this. This looks pretty good, thank you!

**Comment by s1monw** [comment]
@elasticmachine ok to test

**Comment by s1monw** [comment]
@clandry94 can you merge your branch up to current master please

**Comment by s1monw** [comment]
@clandry94 there are some test failures, would you mind taking a look?

**Comment by clandry94** [comment]
Sure, I'll resolve them

**Comment by s1monw** [comment]
@clandry94 you don't have to forcepush for all changes it's easier to review if you don't. you can still just do a `git merge master` if you wanna catch up. we will do a squash and merge once we are ready here

**Comment by clandry94** [comment]
@s1monw in the elasticsearch-ci/2 `IndexStatsMonitoringDocTests.testToXContent` is failing, but for reasons that seem unrelated to the changes I made. Specifically, 
expected:
`{"query_total":18,"query_time_in_millis":19}`
and actual:
`{"query_total":19,"query_time_in_millis":20}`

The actual values seem to be consistent over multiple tests, so should the expected values be changed to match that?

**Comment by s1monw** [comment]
Sorry for the late reply.

> The actual values seem to be consistent over multiple tests, so should the expected values be changed to match that?

I think this change is the reason for this off-by-one: https://github.com/elastic/elasticsearch/pull/38643/files#diff-07c9dc27cfd5ff0ef6a21088ddf8b8d6R327

**Comment by s1monw** [comment]
you also need to merge with master since we updated to a new and incompatible lucene snapshot.

**Comment by dnhatn** [comment]
@clandry94 I tried to get this ready, but I did not have your permission. Could you please merge the master branch to your branch and apply this patch [38643.txt](https://github.com/elastic/elasticsearch/files/2972874/38643.txt)? We need to handle BWC. Thank you!

**Comment by clandry94** [comment]
Hey @dnhatn, merged master and applied your patch. Thanks for the help 😁

**Comment by dnhatn** [comment]
run elasticsearch-ci/packaging-sample

**Comment by dnhatn** [comment]
Hey @clandry94, can you please merge master into your branch? BWC was disabled in the revision that you merged. Thank you!

**Comment by dnhatn** [comment]
Thanks @clandry94 for working on this.

**Pull request merged**